pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+const (
+	VirtualAPIExportMaximumPermissionPolicyAuditPrefix   = "virtual.apiexport.maxpermissionpolicy.authorization.kcp.dev/"
+	VirtualAPIExportMaximumPermissionPolicyAuditDecision = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "decision"
+	VirtualAPIExportMaximumPermissionPolicyAuditReason   = VirtualAPIExportMaximumPermissionPolicyAuditPrefix + "reason"
+)
+
+const notEnforcedMaximumPermissionPolicyReason = "claimed resource does not enforce maximal permission policy"
+
+type maximalPermissionAuthorizer struct {
+	getAPIExport                        func(clusterName, apiExportName string) (*apisv1alpha1.APIExport, error)
+	newDeepSARAuthorizer                func(clusterName logicalcluster.Name) (authorizer.Authorizer, error)
+	getAPIExportByIdentity              func(identityHash string) (*apisv1alpha1.APIExport, error)
+	getAPIBindingReferenceForAttributes func(attr authorizer.Attributes, clusterName logicalcluster.Name) (ref *apisv1alpha1.ExportReference, found bool, err error)
+}
+
+// NewMaximalPermissionAuthorizer creates an authorizer that checks the maximal permission policy
+// for the requested resource if the resource is a claimed resource in the requested API export.
+// The check is omitted if the requested resource itself is not associated with another API export.
+//
+// If the request is a cluster request the authorizer skips authorization if the request is not for a bounded resource.
+// If the request is a wildcard request this check is skipped because no unique API binding can be determined.
+func NewMaximalPermissionAuthorizer(deepSARClient kubernetes.ClusterInterface, apiExportInformer apisinformers.APIExportInformer, apiBindingInformer apisinformers.APIBindingInformer) authorizer.Authorizer {
+	apiExportLister := apiExportInformer.Lister()
+	apiExportIndexer := apiExportInformer.Informer().GetIndexer()
+	apiBindingIndexer := apiBindingInformer.Informer().GetIndexer()
+	indexers.AddIfNotPresentOrDie(apiBindingIndexer, indexers.ClusterScoped())
+	indexers.AddIfNotPresentOrDie(apiExportIndexer, indexers.ClusterScoped())
+
+	return &maximalPermissionAuthorizer{
+		getAPIExport: func(clusterName, apiExportName string) (*apisv1alpha1.APIExport, error) {
+			return apiExportLister.Get(clusters.ToClusterAwareKey(logicalcluster.New(clusterName), apiExportName))
+		},
+		getAPIExportByIdentity: func(identityHash string) (*apisv1alpha1.APIExport, error) {
+			exports, err := apiExportIndexer.ByIndex(indexers.APIExportByIdentity, identityHash)
+			if err != nil {
+				return nil, err
+			}
+			if len(exports) == 0 {
+				return nil, errors.New("export not found")
+			}
+			return exports[0].(*apisv1alpha1.APIExport), nil
+		},
+		newDeepSARAuthorizer: func(clusterName logicalcluster.Name) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(clusterName, deepSARClient)
```

Comment:
What do we gain by factoring this out?

Reply:
same as https://github.com/kcp-dev/kcp/pull/1882#discussion_r966004349